File: src/leetcode/Easy/680.验证回文串-ii.js
/*
 * @lc app=leetcode.cn id=680 lang=javascript
 *
 * [680] 验证回文串 II
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var validPalindrome = function (s) {
  let left = 0
  let right = s.length - 1

  var isPalindrome = (left, right) => {
    for (let i = left, j = right; i < j; i++, j--) {
      if (s[i] !== s[j]) return false
    }
    return true
  }

  while (left < right) {
    if (s[left] === s[right]) {
      left++
      right--
    } else {
      return isPalindrome(left + 1, right) || isPalindrome(left, right - 1)
    }
  }

  return true
}

// @lc code=end

File: src/leetcode/rust/src/easy/680.验证回文串-ii.rs
pub fn valid_palindrome(s: String) -> bool {
    let s_bytes = s.as_bytes();
    let (mut left, mut right) = (0, s.len() - 1);

    let is_palindrome = |left: usize, right: usize| {
        let (mut i, mut j) = (left, right);

        while i < j {
            if s_bytes[left] != s_bytes[right] {
                return false;
            }

            i += 1;
            j -= 1;
        }

        true
    };

    while left < right {
        if s_bytes[left] == s_bytes[right] {
            left += 1;
            right -= 1;
        } else {
            return is_palindrome(left + 1, right) || is_palindrome(left, right - 1);
        }
    }

    true
}
